### ISSUE TITLE
[Future Feature] - Any plans to support custom prefab handout to clients?

### ISSUE THREAD

**Issue by sbtoonz** [PeriodicSeizures/Valhalla#2]
First of all I would love to say amazing project I am EXTREMELY impressed 
Secondly

Is there any planned integrations in order to allow a client to either 

1. Load a list of prefabs ( assetbundle network load async) 
    * if this list of prefabs is loaded server side would allow clients to use mods that have custom prefabs etc
    *  If the server cant load the asset bundles from the clients via an API call.. Can the server hold the assetbundles from mods in order to instantiate XYZ prefab to all players 
    
 2.  Can this server handle storing of cfg files, a huge use case of the mods on a server is to allow the server to hold the mods settings in order to hand them out to the players at connection time. So that the players might not fiddle with potentially game breaking settings as deemed by the mod author at time of creation
 
 
 Again amazing project. Hopefully I can get myself up to speed on your code base and give some contributions

**Comment by crazicrafter1** [comment]
From a glance, it doesn't appear that Jotunn supports raw as-is Unity prefabs being distributed across the network. The easier approach would be to enforce each client to have the same mods on the same version. Jotunn does support this. 

The Jotunn config-sync mechanism would have to be implemented either in C++ or Lua. I do not want the primary C++ codebase to become flooded with non-core functionality, so I'm thinking the config-sync machine would exist in the Lua mod environment as a standalone API mod.

Knowing how the server has very little actual mod interaction (99% is handled by clients, the server is just a middle-man and initial config sync mechanism, and maybe handles misc things), implementing each specific mod should be straightforward and *hopefully* minimal. 

Valheim mod-specific configs (assuming they are in the same format) should be able to be used as-is without migration or any changes being made to function with the Valhalla-server mod. 

Thanks for your feedback! I will be trying to implement the above in the coming weeks.

correction: 
It doesn't seem that Jotunn has the config synchronization? It took a bit of searching, but looking through several Valheim mods, all the ones making use of config-sync all include a packaged ServerSync, which ended up being this https://github.com/blaxxun-boop/ServerSync/blob/master/ConfigSync.cs.

**Comment by sbtoonz** [comment]
Yep am fairly well versed in the mod community 
Jotunn brought in config sync via blaxxun


I make heavy use of serversync's customconfig enforcement in 1 of my mods and use its basic config values in all of my other mods essentially. It is a staple of the modding community at this point in its lifecycle

I realistically was hopeful for a networkload of assetbundle so said client could inform server of held assets and then ensure all connected clients had same list.


I was thinking server would need to know what the prefab was if it was to do any of the spawn logic, 

The "games" way of instantiation to ensure clients see the prefab done through an RPC propigated to all connected clients

the following code is from the game files:

```
  public void SpawnObject(Vector3 pos, Quaternion rot, GameObject prefab)
  {
    int prefabHash = this.GetPrefabHash(prefab);
    ZRoutedRpc.instance.InvokeRoutedRPC(ZRoutedRpc.Everybody, nameof (SpawnObject), (object) pos, (object) rot, (object) prefabHash);
  }
```

That doesnt stop just plain old Object.Instantiate() from working either due to the way the Znet* code was setup where the zdo is set with the prefab name itself

```
else
      {
        string prefabName = this.GetPrefabName();
        this.m_zdo = ZDOMan.instance.CreateNewZDO(this.transform.position);
        this.m_zdo.m_persistent = this.m_persistent;
        this.m_zdo.m_type = this.m_type;
        this.m_zdo.m_distant = this.m_distant;
        this.m_zdo.SetPrefab(prefabName.GetStableHashCode());
        this.m_zdo.SetRotation(this.transform.rotation);
```

That is more or less why I was wondering about the ability to run custom assets on the server itself as from my experience the asset needs to live on the server for all players to see the VFX from weapons VFX from monsters . I am unsure about pieces as I didnt make many piece related mods..

**Comment by crazicrafter1** [comment]
> I realistically was hopeful for a networkload of assetbundle so said client could inform server of held assets and then ensure all connected clients had same list.

I did some digging. There is nowhere within the blaxxun code that Assets are verified across the network. I checked PotionsPlus and RareMagicPortal and I can't find anything resembling this. There is one thing that should take care of this, and that is the VersionCheck in ServerSync. 

Think about it— if all connected clients pass the mod check during the handshake (meaning they are all using the same mods, on a commonly supported version), then they **MUST** be using the same assets because the mods are all the same.

> The "games" way of instantiation to ensure clients see the prefab done through an RPC propigated to all connected clients
> 
> the following code is from the game files:
> 
> ```
>   public void SpawnObject(Vector3 pos, Quaternion rot, GameObject prefab)
>   {
>     int prefabHash = this.GetPrefabHash(prefab);
>     ZRoutedRpc.instance.InvokeRoutedRPC(ZRoutedRpc.Everybody, nameof (SpawnObject), (object) pos, (object) rot, (object) prefabHash);
>   }
> ```
`ZNetScene.SpawnObject` is for bonfire fireworks (and maybe some mods use it?). It can be ignored 99.99% of the time until the devs decide to do more with it.
> 
> That doesnt stop just plain old Object.Instantiate() from working either due to the way the Znet* code was setup where the zdo is set with the prefab name itself.
> 
> That is more or less why I was wondering about the ability to run custom assets on the server itself as from my experience the asset needs to live on the server for all players to see the VFX from weapons VFX from monsters . I am unsure about pieces as I didnt make many piece related mods..

Within ZNetScene, if no prefab can be matched to the ZDOs prefab hash, the ZDO is deleted. This also carries true in Valhalla, but with some exceptions. 

Since there is no equivalent ZNetScene, it does the ZDO validation immediately during `ZDOManager::ZDOData`. I could disable the Prefab enforcement (which would break the nonnull prefab guarantee in the Lua API). I suppose it's up to the end-user anyway so why not...

This is gonna be a lot of work but there's plenty of time.

**Comment by sbtoonz** [comment]
> > I realistically was hopeful for a networkload of assetbundle so said client could inform server of held assets and then ensure all connected clients had same list.


^^ was definitely just a pipedream. 

I do really appreciate the replies and activity.  I was hoping to be able to offer more than a wishlist (pull req or similar ) maybe in the coming weeks I can offer something like this. 


I had started digging through the code earlier today and getting a general grasp of how it all runs. 



Update:

It does look like disabling null check server side would allow clients to spawn custom prefab 

I added a custom enemy (this is done by patching znetscene awake and inserting to m_prefabs and then patching zonesytem for spawners) ... walked by the spawner and was disconnected on the attempt to instantiate the prefab that is custom

```
←[33m[09:16:58] [main thread/WARNING]: Peer error←[0m
←[33m[09:16:58] [main thread/WARNING]: prefab not found←[0m
[09:16:58] [main thread/INFO]: NetConnectionStatusChanged: None, old: Connected
[09:16:58] [main thread/INFO]: Cleaning up peer
```

server doesnt hold that prefab technically and theres no way for it to know the client holds the prefab so it just tossed that zdo and killed my connection.

I will put together a more detailed report forgive the grade school level report above
